content/advent-2018/how-to-send-and-receive-sms.md

```diff
@@ -0,0 +1,501 @@
+Once a `delivery notification operation` message is read, it sends an acknowledgement response packet to the SMSC.
+```go
+// Read deliver notifications from deliverNotifCh channel. 
+func readDeliveryNotif(/*....*/) {
+  wg.Add(1)
+  go func() {
+    defer wg.Done()
+    for {
+      select {
+      case <-closeChan:
+        return
+      case dr := <-deliverNotifCh:
+        refNum := dr[refNumIndex]
+        // msg contains the complete delivery status report from the SMSC
+        msg, _ := hex.DecodeString(dr[drMsgIndex]) 
+        // sender is the access code of the SMSC
+        sender := dr[drSenderIndex]
+        // recvr is the mobile number of the recipient subscriber
+        recvr := dr[drRecvrIndex]
+        // scts is the service center time stamp
+        scts := dr[drSctsIndex]
+        msgID := recvr + ":" + scts
+        writer.Write(deliveryNotifAckPacket([]byte(refNum), msgID))
+        writer.Flush()
+      }
+    }
+  }()
+}
+```
+
+### Read Delivery Short Message
+
+To read incoming mobile originating messages, we start the `readDeliveryMsg` goroutine.
+
+```go
+// Reads all delivery short messages(mobile-originating messages) 
+// from the deliverMsgCh channel.
+func readDeliveryMsg(/*....*/) {
+  wg.Add(1)
+  go func() {
+    defer wg.Done()
+      for {
+        select {
+        case <-closeChan:
+          return
+        case mo := <-deliverMsgCh:
+          xser := mo[xserIndex]
+          xserData := parseXser(xser)
+          msg := mo[moMsgIndex]
+          refNum := mo[refNumIndex]
+          sender := mo[moSenderIndex]
+          recvr := mo[moRecvrIndex]
+          scts := mo[moSctsIndex]
+          sysmsg := recvr + ":" + scts
+          msgID := sender + ":" + scts
+
+          // send ack to SMSC with the same reference number
+          writer.Write(deliverySmAckPacket([]byte(refNum), sysmsg))
+          writer.Flush()
+          
+          var incomingMsg deliverMsgPart
+          incomingMsg.sender = sender
+          incomingMsg.receiver = recvr
+          incomingMsg.message = msg
+          incomingMsg.msgID = msgID
+          // further processing    
+        }
+      }
+    }
+  }()
+}
+```
+
+`deliverMsgPart` is a struct that contains the neccessary parts to concatenate and decode the partial incoming mobile originating message.
+
+```go
+// deliverMsgPart represents a deliver sm message part
+type deliverMsgPart struct {
+  currentPart int
+  totalParts  int
+  refNum      int
+  sender      string
+  receiver    string
+  message     string
+  msgID       string
+  dcs         string
+}
+```
+
+
+To handle multi-part mobile originating SMS, we send partial mobile originating messages to `deliverMsgPartCh` channel and complete mobile originating messages to `deliverMsgCompleteCh` channel.
+
+```go
+
+// Reads all deliver sm messages(mobile-originating messages) 
+// from the deliverMsgCh channel.
+func readDeliveryMsg(/*....*/) {
+  wg.Add(1)
+  go func() {
+    defer wg.Done()
+      for {
```

Comment:
I think the indentation here is wrong, doesn't it? Maybe you could double-check if all code snippets in the post were `gofmt`-ed.